platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformWorkspaceIdentity.java

```diff
@@ -17,8 +17,88 @@
+import org.gradle.internal.hash.HashCode;
+import org.gradle.internal.hash.Hasher;
+import org.gradle.internal.hash.Hashing;
 import org.gradle.internal.snapshot.ValueSnapshot;
 
-public interface TransformWorkspaceIdentity extends UnitOfWork.Identity {
-    ValueSnapshot getSecondaryInputsSnapshot();
+class TransformWorkspaceIdentity implements UnitOfWork.Identity {
+    private final ValueSnapshot secondaryInputsSnapshot;
+    private final String uniqueId;
+
+    private TransformWorkspaceIdentity(ValueSnapshot secondaryInputsSnapshot, HashCode uniqueId) {
+        this.uniqueId = uniqueId.toString();
+        this.secondaryInputsSnapshot = secondaryInputsSnapshot;
+    }
+
+    public ValueSnapshot getSecondaryInputsSnapshot() {
+        return secondaryInputsSnapshot;
+    }
+
+    @Override
+    public String getUniqueId() {
+        return uniqueId;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        TransformWorkspaceIdentity that = (TransformWorkspaceIdentity) o;
+
+        return uniqueId.equals(that.uniqueId);
+    }
+
+    @Override
+    public int hashCode() {
+        return uniqueId.hashCode();
+    }
+
+    public static TransformWorkspaceIdentity createMutable(
+        String normalizedInputArtifactPath,
+        String producerBuildTreePath,
+        ValueSnapshot secondaryInputsSnapshot,
+        HashCode dependenciesHash
+    ) {
+        Hasher hasher = Hashing.newHasher();
+        hasher.putString(normalizedInputArtifactPath);
+        hasher.putString(producerBuildTreePath);
+        secondaryInputsSnapshot.appendToHasher(hasher);
+        hasher.putHash(dependenciesHash);
+        return new TransformWorkspaceIdentity(secondaryInputsSnapshot, hasher.hash());
+    }
+
+    public static TransformWorkspaceIdentity createNonNormalizedImmutable(
+        ValueSnapshot inputArtifactPath,
+        ValueSnapshot inputArtifactSnapshot,
+        ValueSnapshot secondaryInputsSnapshot,
+        HashCode dependenciesHash
+    ) {
+        Hasher hasher = Hashing.newHasher();
+        inputArtifactPath.appendToHasher(hasher);
+        inputArtifactSnapshot.appendToHasher(hasher);
+        secondaryInputsSnapshot.appendToHasher(hasher);
+        hasher.putHash(dependenciesHash);
+        return new TransformWorkspaceIdentity(secondaryInputsSnapshot, hasher.hash());
+    }
+
+    public static TransformWorkspaceIdentity createNormalizedImmutable(
+        ValueSnapshot inputArtifactPath,
+        CurrentFileCollectionFingerprint inputArtifactFingerprint,
+        ValueSnapshot secondaryInputsSnapshot,
+        HashCode dependenciesHash
+    ) {
+        Hasher hasher = Hashing.newHasher();
+        inputArtifactPath.appendToHasher(hasher);
+        hasher.putHash(inputArtifactFingerprint.getHash());
```

Comment:
🎉So we use normalized for projects, nice.

I guess here performance regression are not seen, since project normally doesn't have a lot of project dependencies, right?

Reply:
I'm not sure I understand your question about performance. We already used normalized for project artifacts (both incremental and non-incremental). And why would the number of dependencies matter for calculating a hash for the primary artifact?

Reply:
In this spike https://github.com/gradle/gradle/pull/26765 we did normalization for all artifacts, so we did see some performance regression. But quoted sentence explains my question why we don't see regression here:
>  We already used normalized for project artifacts (both incremental and non-incremental)

---

> And why would the number of dependencies matter for calculating a hash for the primary artifact?

I should phrase it differently: my assumption was that in dependency graph there is less project dependencies then external dependencies, and that is why we don't see a regression. But my assumption was incorrect and we don't see any regression because we anyway did a normalization already before.

Reply:
Yep, that's correct.

Reply:
> We already used normalized for project artifacts (both incremental and non-incremental).
I don't understand this. For project artifacts, we used a mutable workspace, so we didn't normalize the contents for the workspace location. And now instead, we put non-incremental project transforms into a normalized immutable workspace location. We did normalize the inputs of project transforms in every build, since they have not been a part of the cross-build identity cache.

I think the main point here is that we didn't change anything for external dependencies regarding their workspaces, so this extra work doesn't show up.